### DISCUSSION TITLE
Function ID database

### DISCUSSION THREAD

**Discussion by atomicdog** [NationalSecurityAgency/ghidra#3989] [Q&A] [answered]
I cant seem to populate a _fidb_, It always silently fails (FID Dbviewer is empty). There's no indication of anything going wrong, so what can I do to troubleshoot the issue?

I'm trying to make a _fidb_ using object files from FreeRTOS lib compiled for cortex-m4.  I can hash a single function using the script (_Shift+H_) though.

I'm following the direction in the help file and I'm not sure what I'm doing wrong.

**Comment by ghidra1** [comment, marked as answer]
Do you get any errors when you run the _Populate FidDb from programs..._ action?  You should see a results popup dialog if anything gets ingested.  You will need to post your step-by-step actions for us to help.   

I noticed that if the ingest encounters any program within the selected project folder, or its subfolders, that needs to be upgraded due to a Ghidra version change, the fidb population task will terminate immediately and not save any of its results.  If this occurs you will get a popup indicating the details of a VersionException.  I would think this should simply skip the offending file or force an temporary upgrade to allow it to be processed. 

I also noticed that when prompted to select a Processor Language you are actually selecting both a Language ID and Compiler Spec ID even though the settings dialog only shows the Language ID.  Both of these must exactly match the program files being processed and will skip those that do not match both criteria.

There is certainly much room for improvement in how the files are getting scanned and is rather error prone.

**Reply by atomicdog** [reply]
Thank you. 
The issue was the language ID not matching exactly. 

There was no errors or indications of anything going wrong when trying to populate the FidDb (with mismatched Language ID's).

**Reply by astrelsky** [reply]
Just a thought, it would be nice to be able to have a "loader" (or importer I don't remember the correct interface) for a and lib files that would make an fidb for you. Would simplify things.

**Reply by dlenski** [reply]
Is there any documentation on how to do this?

I'm trying to create a FidDB database for the mbedtls libraries compiled for ESP32 (https://github.com/esp-rs/esp-mbedtls/tree/main/esp-mbedtls-sys/libs/xtensa-esp32-none-elf) and just flailing around with no success.

**Reply by gemesa** [reply]
> Is there any documentation on how to do this?
> 
> I'm trying to create a FidDB database for the mbedtls libraries compiled for ESP32 (https://github.com/esp-rs/esp-mbedtls/tree/main/esp-mbedtls-sys/libs/xtensa-esp32-none-elf) and just flailing around with no success.

https://github.com/NationalSecurityAgency/ghidra/blob/master/DevGuide.md#building-fid-databases
https://github.com/NationalSecurityAgency/ghidra/pull/7127/files

**Reply by dlenski** [reply]
Thanks @gemesa. I've now managed to _create_ the FidDB database, but it doesn't seem to _match_ anything and I've been having trouble getting feedback on _why_ not: https://github.com/NationalSecurityAgency/ghidra/discussions/7515
